### ISSUE TITLE
Member registration to Blockchain Bluemix no longer work after a month?

### ISSUE THREAD

**Issue by git123456** [IBM-Blockchain/ibm-blockchain-issues#31]
Hope this is the right place to post my specific blockchain question... different people told me different places to post questions... I am confuse.

Anyway, here's my question, what made a simple new user registration program to register a new user to BlkChain in BlueMix works a month ago but no longer work anymore?  There is no change in  us.blockchain.ibm.com.cert and registra  password( WebAppAdmin). Part of the error statements as shown below :   var signKey = self.cryptoPrimitives.ecdsaKeyFromPrivate(registrar.getEnrollment().key, 'hex');
                                                                                        ^
TypeError: Cannot read property 'key' of undefined
   at MemberServicesImpl.register (/Users/ibm/Documents/workspace/IGF/fabric/node_modules/hfc/lib/hfc.js:1840:90)
   at Member.register (/Users/ibm/Documents/workspace/IGF/fabric/node_modules/hfc/lib/hfc.js:688:29)
   at Member.registerAndEnroll (/Users/ibm/Documents/workspace/IGF/fabric/node_modules/hfc/lib/hfc.js:747:14)
.....    Any idea?

**Comment by angrbrd** [comment]
Hi @git123456 

Can you please clarify what version of hfc module you are using? Given that you are using it with Bluemix, you should use version 0.5.3.

Thanks!

**Comment by git123456** [comment]
I was desperate... I tried 0.5.0 and then 0.6.0, all failed (wasted a few days debugging this)

Is there a page in blumix to specify what version of hfc one should use?

**Comment by git123456** [comment]
ok, I give 0.5.3 a try.  again trying to register a new user (chain_test), used us.blockchain.ibm.com.cert, dev mode, below is part of the logs

# Enroll WebAppAdmin

ok 1 Successfully got WebAppAdmin member
Tue, 13 Sep 2016 13:48:05 GMT hfc Enrolling [req={"enrollmentID":"WebAppAdmin","enrollmentSecret":"fd95d1bd02"}]
Tue, 13 Sep 2016 13:48:05 GMT hfc [MemberServicesImpl.enroll] [{"enrollmentID":"WebAppAdmin","enrollmentSecret":"fd95d1bd02"}]
Tue, 13 Sep 2016 13:48:05 GMT hfc [MemberServicesImpl.enroll] Generating keys...
Tue, 13 Sep 2016 13:48:05 GMT hfc [MemberServicesImpl.enroll] Generating keys...done!
Tue, 13 Sep 2016 13:48:05 GMT hfc [MemberServicesImpl.enroll] Generating request! "3059301306072a8648ce3d020106082a8648ce3d03010703420004c75865420cb54b2fbd9446bc591867db5469a880d9d90e09836428a10d32b1d910b4c44b17a606a7878d35099c3e623ae2ea5c10d2d10d8448698f505958ccbd"
Tue, 13 Sep 2016 13:48:05 GMT hfc [MemberServicesImpl.enroll] Adding signing key!
Tue, 13 Sep 2016 13:48:05 GMT hfc [MemberServicesImpl.enroll] Assding encryption key!
Tue, 13 Sep 2016 13:48:05 GMT hfc [MemberServicesImpl.enroll] [Contact ECA] {"ts":{"seconds":{"low":1473774485,"high":0,"unsigned":false},"nanos":0},"id":{"id":"WebAppAdmin"},"tok":{"tok":{"buffer":{"type":"Buffer","data":[102,100,57,53,100,49,98,100,48,50]},"offset":0,"markedOffset":-1,"limit":10,"littleEndian":false,"noAssert":false}},"sign":{"type":0,"key":{"buffer":{"type":"Buffer","data":[48,89,48,19,6,7,42,134,72,206,61,2,1,6,8,42,134,72,206,61,3,1,7,3,66,0,4,199,88,101,66,12,181,75,47,189,148,70,188,89,24,103,219,84,105,168,128,217,217,14,9,131,100,40,161,13,50,177,217,16,180,196,75,23,166,6,167,135,141,53,9,156,62,98,58,226,234,92,16,210,209,13,132,72,105,143,80,89,88,204,189]},"offset":0,"markedOffset":-1,"limit":91,"littleEndian":false,"noAssert":false}},"enc":{"type":0,"key":{"buffer":{"type":"Buffer","data":[48,89,48,19,6,7,42,134,72,206,61,2,1,6,8,42,134,72,206,61,3,1,7,3,66,0,4,108,34,45,39,162,94,41,41,121,20,135,165,231,108,223,88,223,91,94,161,147,91,14,144,121,35,212,217,14,30,228,22,105,114,167,69,35,25,19,10,27,35,214,190,32,198,178,76,235,152,123,213,158,13,175,231,36,114,41,219,163,155,184,78]},"offset":0,"markedOffset":-1,"limit":91,"littleEndian":false,"noAssert":false}},"sig":null} 
Tue, 13 Sep 2016 13:48:06 GMT hfc [MemberServicesImpl.enroll] failed to create cert pair: err={"code":2,"metadata":{"_internal_repr":{}}}
Tue, 13 Sep 2016 13:48:06 GMT hfc [memberServices.enroll] err=Error: Identity or token does not match., enrollment=undefined
not ok 2 Failed to enroll WebAppAdmin member ---> Error: Identity or token does not match.

**Comment by mrshah-at-work** [comment]
Looks like you already registered with the user once and then deleted the credentials that you received from the membership services. These credentials are stored in the keyvalue store ( @angrbrd correct me if I am wrong).
Once these credentials are registered with the membership services, you cannot reuse the same enrollId. It will error `Identity or token does not match`

If you are using bluemix to host your app, I can quickly think of the following 2 options:
1. (Make admin certs part of the app): Instead of using `/var/keyValueStore` or `/tmp/var/keyValueStore` folder, you can use `__dirname + '/keyValueStore'`. This will save all the keyValues in the current folder.
How does this help? Run the app on local machine, this will download the admin certs to the given folder and then `cf push` will push the it to bluemix.
Then you will not have trouble registering admin user as you will already have the certs any time bluemix redeploys the app. The admin user gives you the ability to add more users to membership services.
1. (Use cloudant or object storage service): Use some service to store the enrollment certs, so they persist between redeployment of the apps.

Hope this helps.

**Comment by git123456** [comment]
Yes you were right webadmin user was registered before, the reason I tried to register it again was because credential which was stored locally was deleted while I was trying to debug the SSL problem (hfc) 0.6.0, issue 32,  with my docker image hence I wanted to register again so that I am able to download the credential.  How do get around this?  Is there a way to clear the existing webadmin registration and regsiter again in bluemix ?  OR is there a way to download the credential from bluemix?  I don't understand what you meant by "run the app on local machine"?  Run local fabric network?  If so then how could both certificates be usable in 2 different fabric network?  But I am still not able to run my docker image because I am stuck with the SSL problem which I have posted here, issue #32

**Comment by git123456** [comment]
... more updates... interesting thing is I tried the online swagger api in bluemix, I was able to log-in using all the existing users, for examples: admin, user_type1_6f89233816, etc.. except having problem to log-in as WebAppAdmin, the error returned "id or token mismatch".  Any idea?

**Comment by mrshah-at-work** [comment]
As you commented above, the webadmin user was registered before and thus you cannot re register it.

By running the app locally, I meant the _nodeJS app_. I thought you were running the nodeJS app on bluemix and when you redeploy, you were losing the certificates. Sorry for the misunderstanding.

So when you register a user with membership services, it returns the enrollment certificate for the user and then scrambles the enrollment information of the user. Thus, you cannot re-register the user and it says that `identity or token mismatch`

**Comment by git123456** [comment]
Hi mrchah,

Note that I did try running it locally and  online swagger api in bluemix just trying to find out what happen to WebAppAdmin account.  Now let's forget trying to registere WebAppAdmin again, do read my other notes which I will paste it here "... more updates... interesting thing is I tried the online swagger api in bluemix, I was able to log-in using all the existing users, for examples: admin, user_type1_6f89233816, etc.. except having problem to log-in as WebAppAdmin, the error returned "id or token mismatch". Any idea?"

so basically I still have problem with WebAppAdmin account.

**Comment by dshuffma-ibm** [comment]
the issue is in the differences in how each account was actually registered.
- webappadmin was registered using HFC using gRPC, and its key value store got lost, this prevents it from ever being used again
- the other enroll IDs were registered using HTTP.  Due to underlining mechanisms of how the HTTP APIs were built you could think of their key value store as on the peer.  Therefore their certs are not lost. they can be used again and again.

**Comment by git123456** [comment]
Dshuffma
- if the key value is lost then is there's a way to regenerate it or clear it?  If not then does it mean this fabric network basically useless now?

I provision a new blockchain in bluemix, tried the online swagger api, have no problem to register webappadmin but same error when tried to log-in as webappadmin.

**Comment by dshuffma-ibm** [comment]
there is no way to recover. yep its useless, well from that ID, which is a pretty important ID.

You will not be able to intermix the two registering methods.  if you use the swagger api on 1 ID, you are locked into always using it, its not going to work via HFC anymore.

Your best bet is to pretend the HTTP APIs don't exist and use HFC to do everything. The HTTP APIs really are only there to help someone really quickly explore the network.  Like I have 3 minutes to spare let me push some buttons and see what happens type of user.  If you are going to build anything then use gRPC aka use the HFC.

**Comment by angrbrd** [comment]
@git123456, furthermore hfc will become the only supported method going forward. So I completely agree with @dshuffma-ibm with using hfc only to build your apps.

**Comment by git123456** [comment]
Waah this is such a crucial information which I don't believe I have read it or was informed before... I really think this should be somehow make known so that no one has to go thru what I have gone thru, basically wasted about 2-3 days on this.

**Comment by dshuffma-ibm** [comment]
Sorry about that, and I agree with you.  At which point do you feel it would have been the most helpful to read this information? Is it something that should be on the swagger page itself? Should it be in the HFC documentation? Both?

**Comment by git123456** [comment]
dshuffma, yes I think it should be in both places seem things are not really centralized yet .. this information should be bolded, flashes..etc so that no one would miss it.

**Comment by git123456** [comment]
This is issue can be closed now.  I have decided to delete the old bluemix blkchain and create a new one, use the right version of hfc and only use hfc and was able to register webappadmin and got the certificate created locally.

**Comment by dshuffma-ibm** [comment]
great, thanks for the feedback and good luck haha

**Comment by angrbrd** [comment]
@git123456, we do appreciate your feedback very much and want to improve this process. I tried to update the current npm README to give more info as far as compatibility goes. But if you could let me know how else I can make it more clear for you to understand, etc. I would greatly appreciate it. If you have a list of some specific suggestions, please let me know. Thanks.
